catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/ws/JsonRpc.java

```diff
@@ -97,83 +100,126 @@ public void onClose(Session session, int statusCode, String reason) {
 
   @Override
   public void onError(Session session, Throwable ex) {
-    // no action required on error
+    if (ex instanceof SecureWebSocketException) {
+      SecurityLogger.audit("Received WebSockets request for user that is not logged in.");
+      handleMessage(
+          session,
+          ((SecureWebSocketException) ex).getWsMessage(),
+          (message, id) -> error(NOT_LOGGED_IN_ERROR, ex.getMessage()));
+    } else {
+      // no action required
+    }
   }
 
-  private Object exec(Object message) {
+  @Override
+  public void onMessage(Session session, String message) throws IOException {
+    handleMessage(session, message, this::callMethod);
+  }
 
-    if (!(message instanceof Map)) {
-      return response(null, invalid("message must be a map", message));
+  private void handleMessage(
+      Session session, String message, BiFunction<Map, Object, Object> handleFunc) {
+    Object id;
+    Object result;
+
+    try {
+      Map messageMap = parseMessage(message);
+      id = parseId(messageMap);
+      validateJsonRpcVersion(messageMap, id);
+      result = handleFunc.apply(messageMap, id);
+    } catch (JsonRpcException exception) {
+      id = exception.messageId;
+      result = exception.error;
     }
 
-    Map msg = (Map) message;
+    session.getRemote().sendStringByFuture(mapper.toJson(response(id, result)));
+  }
+
+  private Map parseMessage(String message) throws JsonRpcException {
+    Object parsed;
 
-    if (!msg.containsKey("id")) {
-      return response(null, invalid("required key `id` missing"));
+    try {
+      parsed = mapper.fromJson(message);
+    } catch (RuntimeException ex) {
+      throw new JsonRpcException(null, error(PARSE_ERROR, "Parse error", message));
+    }
+
+    if (!(parsed instanceof Map)) {
+      throw new JsonRpcException(null, invalid("message must be a map", parsed));
+    }
+
+    return (Map) parsed;
+  }
+
+  private Object parseId(Map message) throws JsonRpcException {
+    if (!message.containsKey("id")) {
+      throw new JsonRpcException(null, invalid("required key `id` missing"));
```

Comment:
What circumstances can cause a message to not have a key "id"?

Reply:
This logic was there before I made any changes. It's all doing validation that's required by the json-rpc specification: https://www.jsonrpc.org/specification